src/Microsoft.AspNetCore.SignalR.Core/StreamTracker.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR.Protocol;
+
+namespace Microsoft.AspNetCore.SignalR
+{
+    internal class StreamTracker
+    {
+        private static readonly MethodInfo _buildConverterMethod = typeof(StreamTracker).GetMethods(BindingFlags.NonPublic | BindingFlags.Static).Single(m => m.Name.Equals("BuildStream"));
+        private ConcurrentDictionary<string, IStreamConverter> _lookup = new ConcurrentDictionary<string, IStreamConverter>();
+
+        /// <summary>
+        /// Creates a new stream and returns the ChannelReader for it as an object.
+        /// </summary>
+        public object AddStream(string streamId, Type itemType)
+        {
+            var newConverter = (IStreamConverter)_buildConverterMethod.MakeGenericMethod(itemType).Invoke(null, Array.Empty<object>());
+            _lookup[streamId] = newConverter;
+            return newConverter.GetReaderAsObject();
+        }
+
+        public Task ProcessItem(StreamDataMessage message)
+        {
+            return _lookup[message.StreamId].WriteToStream(message.Item);
```

Comment:
We should use `TryGetValue` to throw the same error message as we do in `Complete`:

```csharp
    if(_lookup.TryGetValue(message.StreamId, out var converter)
    {
        converter.WriteToStream(message.Item);
    }
    else
    {
        throw new KeyNotFoundException($"No stream with id '{message.StreamId}' could be found.");
    }
```